// tempo.ts/dist/wagmi/Hooks/nonce.js
import { useEffect } from 'react';
import { useChainId, useConfig } from 'wagmi';
import { useQuery } from 'wagmi/query';
import { getNonce, getNonceKeyCount, watchActiveKeyCountChanged, watchNonceIncremented, } from '../Actions/nonce.js';
/**
 * Hook for getting the nonce for an account and nonce key.
 *
 * @example
 * ```tsx
 * import { Hooks } from 'tempo.ts/wagmi'
 *
 * function App() {
 *   const { data, isLoading } = Hooks.nonce.useNonce({
 *     account: '0x...',
 *     nonceKey: 1n,
 *   })
 *
 *   if (isLoading) return <div>Loading...</div>
 *   return <div>Nonce: {data?.toString()}</div>
 * }
 * ```
 *
 * @param parameters - Parameters.
 * @returns Query result with nonce value.
 */
export function useNonce(parameters = {}) {
    const { account, nonceKey, query = {} } = parameters;
    const config = useConfig(parameters);
    const chainId = useChainId({ config });
    const options = getNonce.queryOptions(config, {
        ...parameters,
        chainId: parameters.chainId ?? chainId,
        query: undefined,
    });
    const enabled = Boolean(account && nonceKey !== undefined && (query.enabled ?? true));
    return useQuery({ ...query, ...options, enabled });
}
/**
 * Hook for getting the number of active nonce keys for an account.
 *
 * @example
 * ```tsx
 * import { Hooks } from 'tempo.ts/wagmi'
 *
 * function App() {
 *   const { data, isLoading } = Hooks.nonce.useNonceKeyCount({
 *     account: '0x...',
 *   })
 *
 *   if (isLoading) return <div>Loading...</div>
 *   return <div>Active keys: {data?.toString()}</div>
 * }
 * ```
 *
 * @param parameters - Parameters.
 * @returns Query result with active nonce key count.
 */
export function useNonceKeyCount(parameters = {}) {
    const { account, query = {} } = parameters;
    const config = useConfig(parameters);
    const chainId = useChainId({ config });
    const options = getNonceKeyCount.queryOptions(config, {
        ...parameters,
        chainId: parameters.chainId ?? chainId,
        query: undefined,
    });
    const enabled = Boolean(account && (query.enabled ?? true));
    return useQuery({ ...query, ...options, enabled });
}
/**
 * Hook for watching nonce incremented events.
 *
 * @example
 * ```tsx
 * import { Hooks } from 'tempo.ts/wagmi'
 *
 * function App() {
 *   Hooks.nonce.useWatchNonceIncremented({
 *     onNonceIncremented(args, log) {
 *       console.log('Nonce incremented:', args)
 *     },
 *   })
 *
 *   return <div>Watching for nonce increments...</div>
 * }
 * ```
 *
 * @param parameters - Parameters.
 */
export function useWatchNonceIncremented(parameters = {}) {
    const { enabled = true, onNonceIncremented, ...rest } = parameters;
    const config = useConfig({ config: parameters.config });
    const configChainId = useChainId({ config });
    const chainId = parameters.chainId ?? configChainId;
    useEffect(() => {
        if (!enabled)
            return;
        if (!onNonceIncremented)
            return;
        return watchNonceIncremented(config, {
            ...rest,
            chainId,
            onNonceIncremented,
        });
    }, [config, enabled, onNonceIncremented, chainId, rest]);
}
/**
 * Hook for watching active key count changed events.
 *
 * @example
 * ```tsx
 * import { Hooks } from 'tempo.ts/wagmi'
 *
 * function App() {
 *   Hooks.nonce.useWatchActiveKeyCountChanged({
 *     onActiveKeyCountChanged(args, log) {
 *       console.log('Active key count changed:', args)
 *     },
 *   })
 *
 *   return <div>Watching for active key count changes...</div>
 * }
 * ```
 *
 * @param parameters - Parameters.
 */
export function useWatchActiveKeyCountChanged(parameters = {}) {
    const { enabled = true, onActiveKeyCountChanged, ...rest } = parameters;
    const config = useConfig({ config: parameters.config });
    const configChainId = useChainId({ config });
    const chainId = parameters.chainId ?? configChainId;
    useEffect(() => {
        if (!enabled)
            return;
        if (!onActiveKeyCountChanged)
            return;
        return watchActiveKeyCountChanged(config, {
            ...rest,
            chainId,
            onActiveKeyCountChanged,
        });
    }, [config, enabled, onActiveKeyCountChanged, chainId, rest]);
}
