// tempo.ts/dist/wagmi/Actions/nonce.js
import * as viem_Actions from '../../viem/Actions/nonce.js';
/**
 * Gets the nonce for an account and nonce key.
 *
 * @example
 * ```ts
 * import { createConfig, http } from '@wagmi/core'
 * import { tempo } from 'tempo.ts/chains'
 * import { Actions } from 'tempo.ts/wagmi'
 *
 * const config = createConfig({
 *   chains: [tempo({ feeToken: '0x20c0000000000000000000000000000000000001' })],
 *   transports: {
 *     [tempo.id]: http(),
 *   },
 * })
 *
 * const nonce = await Actions.nonce.getNonce(config, {
 *   account: '0x...',
 *   nonceKey: 1n,
 * })
 * ```
 *
 * @param config - Config.
 * @param parameters - Parameters.
 * @returns The nonce value.
 */
export function getNonce(config, parameters) {
    const { chainId, ...rest } = parameters;
    const client = config.getClient({ chainId });
    return viem_Actions.getNonce(client, rest);
}
(function (getNonce) {
    function queryKey(parameters) {
        return ['getNonce', parameters];
    }
    getNonce.queryKey = queryKey;
    function queryOptions(config, parameters) {
        const { query, ...rest } = parameters;
        return {
            ...query,
            queryKey: queryKey(rest),
            async queryFn({ queryKey }) {
                const [, { account, nonceKey, ...parameters }] = queryKey;
                if (!account)
                    throw new Error('account is required.');
                if (nonceKey === undefined)
                    throw new Error('nonceKey is required.');
                return await getNonce(config, { account, nonceKey, ...parameters });
            },
        };
    }
    getNonce.queryOptions = queryOptions;
})(getNonce || (getNonce = {}));
/**
 * Gets the number of active nonce keys for an account.
 *
 * @example
 * ```ts
 * import { createConfig, http } from '@wagmi/core'
 * import { tempo } from 'tempo.ts/chains'
 * import { Actions } from 'tempo.ts/wagmi'
 *
 * const config = createConfig({
 *   chains: [tempo({ feeToken: '0x20c0000000000000000000000000000000000001' })],
 *   transports: {
 *     [tempo.id]: http(),
 *   },
 * })
 *
 * const count = await Actions.nonce.getNonceKeyCount(config, {
 *   account: '0x...',
 * })
 * ```
 *
 * @param config - Config.
 * @param parameters - Parameters.
 * @returns The number of active nonce keys.
 */
export function getNonceKeyCount(config, parameters) {
    const { chainId, ...rest } = parameters;
    const client = config.getClient({ chainId });
    return viem_Actions.getNonceKeyCount(client, rest);
}
(function (getNonceKeyCount) {
    function queryKey(parameters) {
        return ['getNonceKeyCount', parameters];
    }
    getNonceKeyCount.queryKey = queryKey;
    function queryOptions(config, parameters) {
        const { query, ...rest } = parameters;
        return {
            ...query,
            queryKey: queryKey(rest),
            async queryFn({ queryKey }) {
                const [, { account, ...parameters }] = queryKey;
                if (!account)
                    throw new Error('account is required.');
                return await getNonceKeyCount(config, { account, ...parameters });
            },
        };
    }
    getNonceKeyCount.queryOptions = queryOptions;
})(getNonceKeyCount || (getNonceKeyCount = {}));
/**
 * Watches for nonce incremented events.
 *
 * @example
 * ```ts
 * import { createConfig, http } from '@wagmi/core'
 * import { tempo } from 'tempo.ts/chains'
 * import { Actions } from 'tempo.ts/wagmi'
 *
 * const config = createConfig({
 *   chains: [tempo({ feeToken: '0x20c0000000000000000000000000000000000001' })],
 *   transports: {
 *     [tempo.id]: http(),
 *   },
 * })
 *
 * const unwatch = Actions.nonce.watchNonceIncremented(config, {
 *   onNonceIncremented: (args, log) => {
 *     console.log('Nonce incremented:', args)
 *   },
 * })
 * ```
 *
 * @param config - Config.
 * @param parameters - Parameters.
 * @returns A function to unsubscribe from the event.
 */
export function watchNonceIncremented(config, parameters) {
    const { chainId, ...rest } = parameters;
    const client = config.getClient({ chainId });
    return viem_Actions.watchNonceIncremented(client, rest);
}
/**
 * Watches for active key count changed events.
 *
 * @example
 * ```ts
 * import { createConfig, http } from '@wagmi/core'
 * import { tempo } from 'tempo.ts/chains'
 * import { Actions } from 'tempo.ts/wagmi'
 *
 * const config = createConfig({
 *   chains: [tempo({ feeToken: '0x20c0000000000000000000000000000000000001' })],
 *   transports: {
 *     [tempo.id]: http(),
 *   },
 * })
 *
 * const unwatch = Actions.nonce.watchActiveKeyCountChanged(config, {
 *   onActiveKeyCountChanged: (args, log) => {
 *     console.log('Active key count changed:', args)
 *   },
 * })
 * ```
 *
 * @param config - Config.
 * @param parameters - Parameters.
 * @returns A function to unsubscribe from the event.
 */
export function watchActiveKeyCountChanged(config, parameters) {
    const { chainId, ...rest } = parameters;
    const client = config.getClient({ chainId });
    return viem_Actions.watchActiveKeyCountChanged(client, rest);
}
