// tempo.ts/dist/viem/Storage.js
import { createStore, del, get, set } from 'idb-keyval';
import * as Json from 'ox/Json';
import { normalizeValue } from './internal/utils.js';
export function from(storage, options = {}) {
    const key = (name) => `${options.key ? `${options.key}:` : ''}${name}`;
    return {
        getItem: (name) => storage.getItem(key(name)),
        removeItem: (name) => storage.removeItem(key(name)),
        setItem: (name, value) => storage.setItem(key(name), value),
    };
}
export function idb() {
    const store = typeof indexedDB !== 'undefined'
        ? createStore('tempo.ts', 'store')
        : undefined;
    return from({
        async getItem(name) {
            const value = await get(name, store);
            if (value === null)
                return null;
            return value;
        },
        async removeItem(name) {
            await del(name, store);
        },
        async setItem(name, value) {
            await set(name, normalizeValue(value), store);
        },
    });
}
export function localStorage(options = {}) {
    if (typeof window === 'undefined')
        return memory();
    return from({
        async getItem(name) {
            const item = window.localStorage.getItem(name);
            if (item === null)
                return null;
            try {
                return Json.parse(item);
            }
            catch {
                return null;
            }
        },
        async removeItem(name) {
            window.localStorage.removeItem(name);
        },
        async setItem(name, value) {
            window.localStorage.setItem(name, Json.stringify(value));
        },
    }, options);
}
const store = new Map();
export function memory(options = {}) {
    return from({
        getItem(name) {
            return store.get(name) ?? null;
        },
        removeItem(name) {
            store.delete(name);
        },
        setItem(name, value) {
            store.set(name, value);
        },
    }, options);
}
